betatest/forms.py

```diff
@@ -0,0 +1,199 @@
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Forms for the Instance Manager beta test
+"""
+
+# Imports #####################################################################
+
+from django import forms
+from django.contrib.auth.models import User
+from django.db import transaction
+from django.utils.text import capfirst
+from djng.forms import NgFormValidationMixin, NgModelForm, NgModelFormMixin
+from zxcvbn import password_strength
+
+from betatest.models import BetaTestApplication
+from userprofile.models import UserProfile
+
+
+# Forms #######################################################################
+
+class BetaTestApplicationForm(NgModelFormMixin, NgFormValidationMixin, NgModelForm):
+    """
+    Application form for beta testers. Creates instances of User, UserProfile,
+    and BetaTestApplication models on submit.
+    """
+    class Meta: #pylint: disable=missing-docstring
+        model = BetaTestApplication
+        exclude = ('user', 'status')
+        widgets = {
+            'public_contact_email': forms.widgets.EmailInput(attrs={
+                'validate-email': True,
+            }),
+        }
+
+    full_name = forms.CharField(
+        max_length=255,
+        help_text='Example: Albus Dumbledore',
+    )
+    username = forms.RegexField(
+        regex=r'^[\w.+-]+$',
+        max_length=30,
+        help_text=('This would also be the username of the administrator '
+                   'account on the Open edX instance.'),
+        error_messages={
+            'invalid': ('Usernames may contain only letters, numbers, and '
+                        './+/-/_ characters.'),
+            'unique': 'This username is already taken.',
+        },
+    )
+    email = forms.EmailField(
+        help_text=('This is also your account name, and where we will send '
+                   'important notices.'),
+        widget=forms.widgets.EmailInput(attrs={'validate-email': True}),
+    )
+    password = forms.CharField(
+        widget=forms.PasswordInput,
+        help_text=('Pick a password for your OpenCraft account. You will be '
+                   'able to use it to login and access your account.'),
+    )
+    password_confirmation = forms.CharField(
+        widget=forms.PasswordInput,
+        help_text=('Please use a strong password: avoid common patterns and '
+                   'make it long enough to be difficult to crack.'),
```

Comment:
I agree with Braden. The only problem is that the python port of zxcvbn
does not always give the same results as the js reference implementation.
When Tim was testing this he found a password that passed client-side
validation but failed on the server. It would not be acceptable for a
client-side password strength meter to say that a password is very strong
only for the server to reject it when the form is submitted.

Possible solutions:

- Fix the python port of zxcvbn
- Use pyv8 or similar to call the js version of zxcvbn on the server